# **Income Census Dataset: Initial Exploration**

In [3]:
#import ncessary libraries
import pandas as pd
import numpy as np

In [37]:
# raw data file path
file_path = "../income_dataset/"

In [58]:

# declare the column names test and train files dont have headers.
columns = [
    "age",
    "workclass",
    "fnlwgt",
    "education",
    "education-num",
    "marital-status",
    "occupation",
    "relationship",
    "race",
    "sex",
    "capital-gain",
    "capital-loss",
    "hours-per-week",
    "native-country",
    "income"
]

# load the data
df_train = pd.read_csv(file_path + "adult.data", header=None, names=columns, sep=",",skipinitialspace=True)
df_test = pd.read_csv(file_path + "adult.test", header=None, names=columns, sep=",", skipinitialspace=True, skiprows=1)


In [59]:
df_train.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


In [60]:
df_test.head()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,25,Private,226802,11th,7,Never-married,Machine-op-inspct,Own-child,Black,Male,0,0,40,United-States,<=50K.
1,38,Private,89814,HS-grad,9,Married-civ-spouse,Farming-fishing,Husband,White,Male,0,0,50,United-States,<=50K.
2,28,Local-gov,336951,Assoc-acdm,12,Married-civ-spouse,Protective-serv,Husband,White,Male,0,0,40,United-States,>50K.
3,44,Private,160323,Some-college,10,Married-civ-spouse,Machine-op-inspct,Husband,Black,Male,7688,0,40,United-States,>50K.
4,18,?,103497,Some-college,10,Never-married,?,Own-child,White,Female,0,0,30,United-States,<=50K.


# Explore few rows

In [61]:
df_train.sample()

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
3375,21,Private,305874,Some-college,10,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,54,United-States,<=50K


In [62]:
# get any 2 sample rows
df_train.sample(2, random_state=42)

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
14160,27,Private,160178,Some-college,10,Divorced,Adm-clerical,Not-in-family,White,Female,0,0,38,United-States,<=50K
27048,45,State-gov,50567,HS-grad,9,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K


# Explore columns and data types

In [88]:
#check no of rows and columns
print(f"Train : {df_train.shape}  Test : {df_test.shape}")

Train : (32561, 15)  Test : (16281, 15)


In [63]:
#check datatypes
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [64]:
df_test.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income            object
dtype: object

# check for missing values for both test and train data
 - check missing values using isna() and isnull(). Both fucntions checks same type of missing data
 - check for ? as missing values are represented with ? as per the data document.

In [65]:
df_train.isna().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [66]:
df_test.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64

In [67]:
# check for missing ? counts
(df_train == "?").sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education-num        0
marital-status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital-gain         0
capital-loss         0
hours-per-week       0
native-country     583
income               0
dtype: int64

In [68]:
# get the missing(?) columns for train data
missing_q = (df_train == "?").sum()
missing_q[missing_q > 0].sort_values(ascending=False)

occupation        1843
workclass         1836
native-country     583
dtype: int64

In [69]:
# get the missing(?) columns for test data
missing_t = (df_test == "?").sum()
missing_t[missing_t > 0].sort_values(ascending=False)

occupation        966
workclass         963
native-country    274
dtype: int64

In [72]:
# view the missing values rows 
df_train[df_train.isin(["?"]).any(axis=1)].sample(2)

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
28194,27,?,133359,Bachelors,13,Married-spouse-absent,?,Not-in-family,White,Male,0,0,50,?,<=50K
30822,21,?,105312,HS-grad,9,Never-married,?,Not-in-family,White,Female,0,0,20,United-States,<=50K


# statistical summary of train and test data

In [73]:

df_train.describe(include='all')

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
count,32561.000000,32561,3.256100e+04,32561,32561.000000,32561,32561,32561,32561,32561,32561.000000,32561.000000,32561.000000,32561,32561
unique,NaN,9,NaN,16,NaN,7,15,6,5,2,NaN,NaN,NaN,42,2
top,NaN,Private,NaN,HS-grad,NaN,Married-civ-spouse,Prof-specialty,Husband,White,Male,NaN,NaN,NaN,United-States,<=50K
freq,NaN,22696,NaN,10501,NaN,14976,4140,13193,27816,21790,NaN,NaN,NaN,29170,24720
mean,38.581647,NaN,1.897784e+05,NaN,10.080679,NaN,NaN,NaN,NaN,NaN,1077.648844,87.303830,40.437456,NaN,NaN
std,13.640433,NaN,1.055500e+05,NaN,2.572720,NaN,NaN,NaN,NaN,NaN,7385.292085,402.960219,12.347429,NaN,NaN
min,17.000000,NaN,1.228500e+04,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,1.000000,NaN,NaN
25%,28.000000,NaN,1.178270e+05,NaN,9.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
50%,37.000000,NaN,1.783560e+05,NaN,10.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
75%,48.000000,NaN,2.370510e+05,NaN,12.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,45.000000,NaN,NaN


In [74]:
df_test.describe(include='all')

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
count,16281.000000,16281,1.628100e+04,16281,16281.000000,16281,16281,16281,16281,16281,16281.000000,16281.000000,16281.000000,16281,16281
unique,NaN,9,NaN,16,NaN,7,15,6,5,2,NaN,NaN,NaN,41,2
top,NaN,Private,NaN,HS-grad,NaN,Married-civ-spouse,Prof-specialty,Husband,White,Male,NaN,NaN,NaN,United-States,<=50K.
freq,NaN,11210,NaN,5283,NaN,7403,2032,6523,13946,10860,NaN,NaN,NaN,14662,12435
mean,38.767459,NaN,1.894357e+05,NaN,10.072907,NaN,NaN,NaN,NaN,NaN,1081.905104,87.899269,40.392236,NaN,NaN
std,13.849187,NaN,1.057149e+05,NaN,2.567545,NaN,NaN,NaN,NaN,NaN,7583.935968,403.105286,12.479332,NaN,NaN
min,17.000000,NaN,1.349200e+04,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,1.000000,NaN,NaN
25%,28.000000,NaN,1.167360e+05,NaN,9.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
50%,37.000000,NaN,1.778310e+05,NaN,10.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,40.000000,NaN,NaN
75%,48.000000,NaN,2.383840e+05,NaN,12.000000,NaN,NaN,NaN,NaN,NaN,0.000000,0.000000,45.000000,NaN,NaN


In [75]:
# count of unique values per column in both train and test data
pd.DataFrame({
    'train_nunique': df_train.nunique(),
    'test_nunique': df_test.nunique()
})

,train_nunique,test_nunique
age,73,73
workclass,9,9
fnlwgt,21648,12787
education,16,16
education-num,16,16
marital-status,7,7
occupation,15,15
relationship,6,6
race,5,5
sex,2,2


In [76]:
# check data distribution for numeric features
pd.DataFrame({
    "train_skew" : df_train.select_dtypes(include='number').skew(),
    "test_skew" : df_test.select_dtypes(include='number').skew()
})
#df_train.select_dtypes(include='number').skew()

,train_skew,test_skew
age,0.558743,0.554579
fnlwgt,1.446980,1.422954
education-num,-0.311676,-0.326338
capital-gain,11.953848,11.778293
capital-loss,4.594629,4.520647
hours-per-week,0.227643,0.260419


#### Skewness
- Skewness helps us understand the asymmetry of each numeric feature's distribution.  
- A value close to 0 indicates a roughly symmetric distribution
- positive values show a right‑skew (long right tail)
- negative values show a left‑skew.


# check for Duplicates in train and test data

In [77]:
df_train.duplicated().sum()

24

In [79]:
df_train[df_train.duplicated()].sample(2)

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
30845,46,Private,133616,Some-college,10,Divorced,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
17040,46,Private,173243,HS-grad,9,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,40,United-States,<=50K


In [80]:
df_test.duplicated().sum()

5

In [81]:
df_test[df_test.duplicated()].sample(2)

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
11212,29,Private,36440,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Female,0,0,40,United-States,<=50K.
15960,18,Self-emp-inc,378036,12th,8,Never-married,Farming-fishing,Own-child,White,Male,0,0,10,United-States,<=50K.


---

# **Summary of Initial Exploration**
Train Data : - The dataset has 32,561 rows and 15 columns, containing both numeric and categorical features.
- The categorical columns workclass, occupation, and native‑country contain "?" values that represent missing data.
- There are 24 duplicate rows in the dataset.
- Numeric features are mostly well‑distributed, except capital-gain and capital-loss, which are highly skewed.


Test Data : The test dataset contains 16,281 rows and 15 columns, with the same numeric and categorical structure as the training data.
- The first row is not valid data and should be skipped before analysis
- The categorical columns workclass, occupation, and native‑country contain "?" values that represent missing data.
- There are 5 duplicate rows in the dataset.
- Numeric features follow similar patterns to the train set, with capital-gain and capital-loss remaining highly skewed.



# End of Initial Exploration